### DISCUSSION TITLE
Converting Excel.Range to ExcelReference including SheetId

### DISCUSSION THREAD

**Discussion by emphasis87** [Excel-DNA/ExcelDna#696] [Q&A]
Hi,

I need to compare a large number of ExcelReference instances with a small number of Excel.Range objects, specifically if their SheetId match. 

Is there a way how to get ExcelReference from Excel.Range? 

Previously I was comparing their string representation (xlfReftext and Excel.Range.Address), but that stops working with renaming a sheet leading to all sorts of  complications.

**Comment by emphasis87** [comment]
Looks like I found a way to retrieve SheetId specifically, 

```csharp
Excel.Range range;

var address = range.Address[true, true, Excel.XlReferenceStyle.xlA1, true,null];
var sheetName = rangeId[0..rangeId.LastIndexOf('!')];
var sheetRange = (ExcelRange)XlCall.Excel(XlCall.xlSheetId, sheetName);
var sheetId = sheetRange.SheetId;
```

**Comment by govert** [comment]
I think your way is fine.

There is a constructor of ExcelReference that takes the sheet name:
https://github.com/Excel-DNA/ExcelDna/blob/837305d2fd0eecb016234a736a30dbf250fa9992/Source/ExcelDna.Integration/ExcelReference.cs#L96C1-L103C10

```
        // TODO: Consider how to deal with invalid sheetName. I presume xlSheetId will fail.
        // Perhaps throw a custom exception...?
        public ExcelReference(int rowFirst, int rowLast, int columnFirst, int columnLast, string sheetName)
        {
            ExcelReference sheetRef = (ExcelReference)XlCall.Excel(XlCall.xlSheetId, sheetName);
            this.sheetId = sheetRef.SheetId;
            rectangle = new ExcelRectangle(rowFirst, rowLast, columnFirst, columnLast);
        }
```

To get the name (with the workbook name) I guess you could also try something like `$"'[{myRange.Worksheet.Parent.Name}]{Range.Worksheet.Name}'"`. But I know spaces in workbook names can be handled inconsistently between different APIs, so that's worth checking.
